Add hostspot plugin

IgWod opened the pull request:
This pull requests adds a hotspot plugin. A hotspot is a simple plugin that counts how many times each basic block executes. An example output below:

```
**********************************************************
********************* MAMBO  HOTSTAT *********************
**********************************************************
************** Executing the application... **************
**********************************************************
************** Basic blocks execution count **************
**********************************************************
** Basic block 0x4003e8         executed        1 times **
** Basic block 0x4003f4         executed        1 times **
** Basic block 0x400400         executed        1 times **
** Basic block 0x400420         executed        1 times **
** Basic block 0x400450         executed        1 times **
** Basic block 0x400498         executed        1 times **
** Basic block 0x4004a8         executed        1 times **
** Basic block 0x4004b0         executed        1 times **
** Basic block 0x4004e0         executed        1 times **
** Basic block 0x4004f0         executed        1 times **
** Basic block 0x400520         executed        1 times **
** Basic block 0x400530         executed        1 times **
** Basic block 0x400548         executed        1 times **
** Basic block 0x40054c         executed        1 times **
** Basic block 0x400560         executed        1 times **
** Basic block 0x400570         executed        1 times **
** Basic block 0x4005a0         executed        1 times **
** Basic block 0x4005a8         executed      256 times **
** Basic block 0x4005b4         executed      256 times **
** Basic block 0x4005d0         executed        1 times **
** Basic block 0x4005dc         executed        1 times **
** Basic block 0x4005e8         executed        1 times **
** Basic block 0x4005f8         executed        1 times **
** Basic block 0x400630         executed        1 times **
** Basic block 0x400638         executed        1 times **
** Basic block 0x400654         executed        1 times **
** Basic block 0x40065c         executed        1 times **
** Basic block 0x400674         executed        1 times **
**********************************************************
```

Any comments, please let me know.

lgeek commented on plugins/hotspot.c:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  uint64_t* counter = NULL;
+
```
We don't need separate logic depending on the type of fragment. The same basic block in the application may be scanned multiple times (e.g. because of code cache flushing) so it should always use the code path which attempts a lookup first.

lgeek reviewed (changes requested):
Hi Igor. Thanks for submitting the pull request for the plugin. Ideally, I'd like to improve a few things before merging it, but if you don't have the time for it, let me know:

* the basic block addresses should be printed as an offset relative to a file (e.g.  `0x3e8 from /lib/ld-linux.so` rather than `0x4003e8`). Up to you if you want to replace the absolute address entirely or print it in addition to that.
* where available, the corresponding symbol name should be printed - `get_symbol_info_by_addr()` will do most of the work to address both this and the previous point
* ideally these mappings should be done in the pre_bb event or every time an executable region is unmapped (`mambo_register_vm_op_cb` and look for `VM_UNMAP` events) in case we end up with different executables loaded at overlapping addresses through the lifetime of the application. However that would probably be an uncommon edge case, it would be acceptable to just print a warning about it if the application unmaps files and just do all the lookups when the application exits.
* the basic blocks should be sorted by number of executions, from the highest to the lowest

IgWod commented:
Hi Cosmin! I finally sat and addressed all the feedback (to be honest I completely forgot about this pull request). Ideally we should merge it now, otherwise it may take another few months to fix stuff :) I'm happy to fix any small problems, or obvious bugs that are still left, but if possible I'd like to keep functionality as it is. The plugin on its own is not very useful, so probably we should treat it as an another example on how to use mambo, rather than a full-blown tool. But in any case let me know what you think.

The updated output of the plugin:

```
**********************************************************
********************* MAMBO  HOTSTAT *********************
**********************************************************
************** Executing the application... **************
**********************************************************
Sorted array: 
1 5 7 8 9 10 
We're done; exiting with status: 0
************** Basic blocks execution count **************
**********************************************************
/home/shunya/mambo-1/qs + 0x40 (_Z9partitionPiii) executed 11 times
/home/shunya/mambo-1/qs + 0x54 (_Z9partitionPiii) executed 11 times
/home/shunya/mambo-1/qs + 0xac (_Z9partitionPiii) executed 11 times
/home/shunya/mambo-1/qs + 0x0 (_Z9quickSortPiii) executed 9 times
/home/shunya/mambo-1/qs + 0x68 (_Z9quickSortPiii) executed 9 times
/home/shunya/mambo-1/qs + 0x780 (none) executed 7 times
/home/shunya/mambo-1/qs + 0x750 (none) executed 6 times
/home/shunya/mambo-1/qs + 0x7d0 (none) executed 6 times
/home/shunya/mambo-1/qs + 0x14 (_Z10printArrayPii) executed 6 times
/home/shunya/mambo-1/qs + 0x24 (_Z10printArrayPii) executed 6 times
/home/shunya/mambo-1/qs + 0x44 (_Z10printArrayPii) executed 6 times
/home/shunya/mambo-1/qs + 0x5c (_Z10printArrayPii) executed 6 times
/home/shunya/mambo-1/qs + 0x0 (_Z4swapPiS_) executed 5 times
/home/shunya/mambo-1/qs + 0x0 (_Z9partitionPiii) executed 4 times
/home/shunya/mambo-1/qs + 0xbc (_Z9partitionPiii) executed 4 times
/home/shunya/mambo-1/qs + 0xec (_Z9partitionPiii) executed 4 times
/home/shunya/mambo-1/qs + 0x24 (_Z9quickSortPiii) executed 4 times
/home/shunya/mambo-1/qs + 0x34 (_Z9quickSortPiii) executed 4 times
/home/shunya/mambo-1/qs + 0x50 (_Z9quickSortPiii) executed 4 times
/home/shunya/mambo-1/qs + 0x60 (__libc_csu_init) executed 2 times
/home/shunya/mambo-1/qs + 0x730 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x73c (none) executed 1 times
/home/shunya/mambo-1/qs + 0x770 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x790 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x7a0 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x7b0 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x7f0 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (call_weak_fn) executed 1 times
/home/shunya/mambo-1/qs + 0x10 (call_weak_fn) executed 1 times
/home/shunya/mambo-1/qs + 0x858 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x888 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x898 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x8c8 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x8d8 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x8f0 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x8f4 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x908 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x918 (none) executed 1 times
/home/shunya/mambo-1/qs + 0x74 (_Z9partitionPiii) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (_Z10printArrayPii) executed 1 times
/home/shunya/mambo-1/qs + 0x6c (_Z10printArrayPii) executed 1 times
/home/shunya/mambo-1/qs + 0x80 (_Z10printArrayPii) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (main) executed 1 times
/home/shunya/mambo-1/qs + 0x48 (main) executed 1 times
/home/shunya/mambo-1/qs + 0x5c (main) executed 1 times
/home/shunya/mambo-1/qs + 0x68 (main) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (_Z41__static_initialization_and_destruction_0ii) executed 1 times
/home/shunya/mambo-1/qs + 0x1c (_Z41__static_initialization_and_destruction_0ii) executed 1 times
/home/shunya/mambo-1/qs + 0x2c (_Z41__static_initialization_and_destruction_0ii) executed 1 times
/home/shunya/mambo-1/qs + 0x38 (_Z41__static_initialization_and_destruction_0ii) executed 1 times
/home/shunya/mambo-1/qs + 0x54 (_Z41__static_initialization_and_destruction_0ii) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (_GLOBAL__sub_I__Z4swapPiS_) executed 1 times
/home/shunya/mambo-1/qs + 0x14 (_GLOBAL__sub_I__Z4swapPiS_) executed 1 times
/home/shunya/mambo-1/qs + 0x0 (__libc_csu_init) executed 1 times
/home/shunya/mambo-1/qs + 0x3c (__libc_csu_init) executed 1 times
/home/shunya/mambo-1/qs + 0x40 (__libc_csu_init) executed 1 times
/home/shunya/mambo-1/qs + 0x48 (__libc_csu_init) executed 1 times
/home/shunya/mambo-1/qs + 0x68 (__libc_csu_init) executed 1 times
/home/shunya/mambo-1/qs + 0xcfc (none) executed 1 times
**********************************************************
```
